src/pcnt.rs

```diff
@@ -1,656 +1,1121 @@
-        neg_mode: PcntCountMode,
-        hctrl_mode: PcntControlMode,
-        lctrl_mode: PcntControlMode,
-    ) -> Result<(), EspError> {
-        unsafe {
-            esp!(pcnt_set_mode(
-                self.unit,
-                channel.into(),
-                pos_mode.into(),
-                neg_mode.into(),
-                hctrl_mode.into(),
-                lctrl_mode.into()
-            ))
-        }
+    /// If no pin is specified, a virtual IO pin will be assigned instead.
+    /// The initial level of the virtual IO pin can be configured with the [`ChannelConfig`].
+    ///
+    /// # Number of channels
+    ///
+    /// There is a hardware limit for how many channels a PCNT unit can have.
+    /// For most chips, each PCNT unit can have 2 channels.
+    ///
+    /// # Errors
+    ///
+    /// - `ESP_ERR_INVALID_ARG`: Create PCNT channel failed because of invalid argument
+    /// - `ESP_ERR_NO_MEM`: Create PCNT channel failed because of insufficient memory
+    /// - `ESP_ERR_NOT_FOUND`: Create PCNT channel failed because all PCNT channels are used up and no more free one
+    /// - `ESP_FAIL`: Create PCNT channel failed because of other error
+    pub fn add_channel(
+        &mut self,
+        edge_pin: Option<impl InputPin + 'd>,
+        level_pin: Option<impl InputPin + 'd>,
+        config: &ChannelConfig,
+    ) -> Result<&mut PcntChannel<'d>, EspError> {
+        self.channels
+            .push(unsafe { PcntChannel::new(self.handle, edge_pin, level_pin, config)? })
+            .expect("Not enough channels available.");
+
+        Ok(self.channels.last_mut().unwrap())
     }
 
-    /// Add ISR handler for specified unit.
-    ///
-    /// This ISR handler will be called from an ISR. So there is a stack
-    /// size limit (configurable as \"ISR stack size\" in menuconfig). This
-    /// limit is smaller compared to a global PCNT interrupt handler due
-    /// to the additional level of indirection.
+    /// Set event callback for the PCNT unit.
     ///
     /// # Safety
     ///
-    /// Care should be taken not to call STD, libc or FreeRTOS APIs (except for a few allowed ones)
+    /// Care should be taken not to call std, libc or FreeRTOS APIs (except for a few allowed ones)
     /// in the callback passed to this function, as it is executed in an ISR context.
     ///
-    /// @param callback Interrupt handler function.
+    /// You are not allowed to block, but you are allowed to call FreeRTOS APIs with the FromISR suffix.
+    ///
+    /// # Errors
     ///
-    /// returns
-    /// - ()
-    /// - EspError
+    /// - `ESP_ERR_INVALID_ARG`: Set event callbacks failed because of invalid argument
+    /// - `ESP_FAIL`: Set event callbacks failed because of other error
     #[cfg(feature = "alloc")]
-    pub unsafe fn subscribe<F>(&self, callback: F) -> Result<(), EspError>
-    where
-        F: FnMut(u32) + Send + 'static,
-    {
-        self.internal_subscribe(callback)
+    pub unsafe fn subscribe(
```

Comment:
This does **not** _need_ to be unsafe.

If you - however - had
```rust
on_reach: impl FnMut(WatchEventData) + Send + 'd,
```

then - because of `'d` being non-static then **that** version should've been unsafe. It is unsafe, because 
(a) you need to store then `on_reach` in the box by transmuting `'d` to ` 'static`
(b) if you `core::mem::forget` the driver, and `on_reach` is indeed non-static - booom!

Reply:
But passing a `'d` lifetime has its advantages. That's why all other drivers have two "subscribe" signatures:
- A safe `subscribe` one, where the callback is `'static`
- An unsafe `subscribe_nonstatic` one, where the callback is `'d`. But no core::mem::forget for you in that case

Reply:
Sounds good to me. I thought the unsafe was because the closure is executed in ISR context, I likely confused the two variants